<a href="https://colab.research.google.com/github/nebojsa55/Computational-Genomics_MidTerm-Project/blob/master/notebooks/1.%20Basic-regression-models.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In this notebook we will look at two regression models (**Random forest regressor** and **SVR**) and their respective accuracies when predicting gestational age in pregnant women using gene expression. In the second notebook, we will try to implement ensemble regressor by utilizing two predefined models and using linear regression to get a linear ensemble regressor.

In [1]:
import numpy as np
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

# Navigate to the folder cointaing our data
%cd 'drive/MyDrive/ETF/Master/Computational-Genomics/Project/data'


Mounted at /content/drive
/content/drive/MyDrive/ETF/Master/Computational-Genomics/Project/data


In [2]:
# Verify the chosen folder
!ls

anoSC1_v11_nokey.csv  anoSC2_v20_nokey.csv  eset_HuGene21ST.csv  HTA20_RMA.csv


In [3]:
anno = pd.read_csv('anoSC1_v11_nokey.csv', delimiter = ',', index_col = 0)
anno.head()

GA  Batch      Set  Train Platform
SampleID                                             
Tarca_001_P1A01  11.0      1  PRB_HTA      1    HTA20
Tarca_013_P1B01  15.3      1  PRB_HTA      1    HTA20
Tarca_025_P1C01  21.7      1  PRB_HTA      1    HTA20
Tarca_037_P1D01  26.7      1  PRB_HTA      1    HTA20
Tarca_049_P1E01  31.3      1  PRB_HTA      1    HTA20

In [4]:
HTA20_RMA = pd.read_csv('HTA20_RMA.csv', delimiter = ',', index_col = 0).transpose()
HTA20_RMA.head()

1_at     10_at  ...  AFFX-r2-Ec-bioD-5_at  AFFX-r2-P1-cre-5_at
Tarca_001_P1A01  6.062215  3.796484  ...             12.101476            13.695705
Tarca_003_P1A03  6.125023  3.805305  ...             12.370891            13.635522
Tarca_004_P1A04  5.875502  3.450245  ...             12.295786            13.616688
Tarca_005_P1A05  6.126131  3.628411  ...             12.249936            13.524328
Tarca_006_P1A06  6.146466  3.446812  ...             12.245458            13.509353

[5 rows x 32830 columns]

## Missing data

Unfortunately, all of the test data are missing **GA** (gastetion age) parameter, due to the nature od DREAM Pretem Challenge. In order to overcome this, we will use only the original training set for our model training and testing. 10% of data was used for testing (37 samples). 10-fold cross validation was performed in order to determine our ML model.

In [5]:
anno.loc[anno['Train'] == 1].shape[0]

367

In [6]:
# Sync the X and y data by sorting the labels

df1 = anno.sort_index()
df2 = HTA20_RMA.sort_index()

X = df2.iloc[np.array(np.logical_not(df1['GA'].isna())),:]
y = df1.dropna().loc[:,['GA','Batch']]


In [7]:
# Check to see if the indexes are the same
(X.index == y.index).all()

True

## Standard scaling
We will now perform the standard scaling of features from *one batch*

In [8]:
from sklearn.preprocessing import StandardScaler

XX = np.zeros(X.shape)
for i in [1,2,3,4,5,6,7,8,9,10,32]:
    scale = StandardScaler()
    indices = np.bool8(y['Batch'] == i)
    Xtemp = X.iloc[indices,:]
    scale.fit(Xtemp)
    XX[indices,:] = scale.transform(Xtemp)

In [9]:
# delete batch column
yy = y['GA']

## Regression

In this section we will perform regression using a couple of well known regressors such as:
* **Random forest regressor** and
* **Support vector regressor**

The main regression metric will be **RMSE** (Root mean square error), as it was used in DREAM Preterm challenge. 

The total number of features is 32830, which is a large number. We will shrink the number of features via **PCA analysis** to obtain 95% variance. The PCA will give us a total of /// features.

### Random forest regressor

In [10]:
# import necessary ML modules
from sklearn.decomposition import PCA
from sklearn.model_selection import KFold
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from tqdm import tqdm


num_of_splits = 10
kf = KFold(n_splits = num_of_splits)
regressor = RandomForestRegressor(n_estimators = 1000,
                                  n_jobs = -1, 
                                  random_state = 42)

pca = PCA(n_components = 0.95, svd_solver = 'full')

y_train_pred = []
y_test_pred = []

for train_index, test_index in tqdm(kf.split(XX), total = num_of_splits, unit = 'iteration'):

  X_train_std, X_test_std = XX[train_index,:], XX[test_index,:]
  y_train, y_test = yy.iloc[train_index], yy.iloc[test_index]

  pca.fit(XX)
  
  X_train_pca = pca.transform(X_train_std)
  X_test_pca = pca.transform(X_test_std)
  
  regressor.fit(X_train_pca,y_train)
  
  y_train_pred.append(mean_squared_error(y_train,regressor.predict(X_train_pca), squared = False))
  y_test_pred.append(mean_squared_error(y_test,regressor.predict(X_test_pca), squared = False))



100%|██████████| 10/10 [05:39<00:00, 33.96s/iteration]


In [11]:
np.mean(y_train_pred)

2.7622168255170916

In [12]:
np.mean(y_test_pred)

7.532737936379085

#### Model overfitting
We can observe that our current RF model is heavily overfiting. To overcome that, we will try to tune hyperparameters via **Randomized Search crossvalidation**

In [15]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer
from sklearn.model_selection import RandomizedSearchCV

scorer = make_scorer(mean_squared_error, greater_is_better = False)

n_estimators = [int(x) for x in np.linspace(start = 200, stop = 1000, num = 5)]

# Number of features to consider at every split
max_features = ['auto', 'sqrt']

# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(10, 100, num = 10)]
max_depth.append(None)

# Minimum number of samples required to split a node
min_samples_split = [2, 10]

# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 4]

# Method of selecting samples for training each tree
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
                }


# Use the random grid to search for best hyperparameters
# First create the base model to tune
rf = RandomForestRegressor()
# Random search of parameters, using 3 fold cross validation, 
# search across 100 different combinations, and use all available cores
rf_random = RandomizedSearchCV(estimator = rf, 
                               param_distributions = random_grid, 
                               n_iter = 20, 
                               cv = 5, 
                               verbose = 3, 
                               random_state = 42)

XX_pca = pca.transform(XX)
rf_random.fit(XX_pca,yy)

Fitting 5 folds for each of 20 candidates, totalling 100 fits
[CV] n_estimators=200, min_samples_split=10, min_samples_leaf=1, max_features=sqrt, max_depth=70 


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]  n_estimators=200, min_samples_split=10, min_samples_leaf=1, max_features=sqrt, max_depth=70, score=-0.023, total=   0.7s
[CV] n_estimators=200, min_samples_split=10, min_samples_leaf=1, max_features=sqrt, max_depth=70 


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.7s remaining:    0.0s


[CV]  n_estimators=200, min_samples_split=10, min_samples_leaf=1, max_features=sqrt, max_depth=70, score=0.063, total=   0.6s
[CV] n_estimators=200, min_samples_split=10, min_samples_leaf=1, max_features=sqrt, max_depth=70 


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    1.3s remaining:    0.0s


[CV]  n_estimators=200, min_samples_split=10, min_samples_leaf=1, max_features=sqrt, max_depth=70, score=0.057, total=   0.6s
[CV] n_estimators=200, min_samples_split=10, min_samples_leaf=1, max_features=sqrt, max_depth=70 
[CV]  n_estimators=200, min_samples_split=10, min_samples_leaf=1, max_features=sqrt, max_depth=70, score=0.082, total=   0.6s
[CV] n_estimators=200, min_samples_split=10, min_samples_leaf=1, max_features=sqrt, max_depth=70 
[CV]  n_estimators=200, min_samples_split=10, min_samples_leaf=1, max_features=sqrt, max_depth=70, score=0.059, total=   0.6s
[CV] n_estimators=800, min_samples_split=10, min_samples_leaf=4, max_features=sqrt, max_depth=20 
[CV]  n_estimators=800, min_samples_split=10, min_samples_leaf=4, max_features=sqrt, max_depth=20, score=-0.018, total=   2.3s
[CV] n_estimators=800, min_samples_split=10, min_samples_leaf=4, max_features=sqrt, max_depth=20 
[CV]  n_estimators=800, min_samples_split=10, min_samples_leaf=4, max_features=sqrt, max_depth=20, scor

[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:  8.4min finished


RandomizedSearchCV(cv=5, error_score=nan,
                   estimator=RandomForestRegressor(bootstrap=True,
                                                   ccp_alpha=0.0,
                                                   criterion='mse',
                                                   max_depth=None,
                                                   max_features='auto',
                                                   max_leaf_nodes=None,
                                                   max_samples=None,
                                                   min_impurity_decrease=0.0,
                                                   min_impurity_split=None,
                                                   min_samples_leaf=1,
                                                   min_samples_split=2,
                                                   min_weight_fraction_leaf=0.0,
                                                   n_estimators=100,
                              

In [18]:
rf_random.get_params

<bound method BaseEstimator.get_params of RandomizedSearchCV(cv=5, error_score=nan,
                   estimator=RandomForestRegressor(bootstrap=True,
                                                   ccp_alpha=0.0,
                                                   criterion='mse',
                                                   max_depth=None,
                                                   max_features='auto',
                                                   max_leaf_nodes=None,
                                                   max_samples=None,
                                                   min_impurity_decrease=0.0,
                                                   min_impurity_split=None,
                                                   min_samples_leaf=1,
                                                   min_samples_split=2,
                                                   min_weight_fraction_leaf=0.0,
                                                   n_esti

In [21]:

n_estimators = [100, 1000, 3000]

# Number of features to consider at every split
max_features = ['auto']

# Maximum number of levels in tree
max_depth = [3, None]

# Minimum number of samples required to split a node
min_samples_split = [2, 4]

# Minimum number of samples required at each leaf node
min_samples_leaf = [1,10]

# Method of selecting samples for training each tree
param_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
                }


# Use the random grid to search for best hyperparameters
# First create the base model to tune
rf = RandomForestRegressor()
# Random search of parameters, using 3 fold cross validation, 
# search across 100 different combinations, and use all available cores
rf_grid = GridSearchCV(estimator = rf, 
                       param_grid = param_grid,  
                       cv = 5, 
                       verbose = 3)

rf_grid.fit(XX_pca,yy)

Fitting 5 folds for each of 24 candidates, totalling 120 fits
[CV] max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100 


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]  max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100, score=0.103, total=   1.5s
[CV] max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100 


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    1.5s remaining:    0.0s


[CV]  max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100, score=0.224, total=   1.5s
[CV] max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100 


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    3.0s remaining:    0.0s


[CV]  max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100, score=0.189, total=   1.5s
[CV] max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100 
[CV]  max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100, score=0.262, total=   1.4s
[CV] max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100 
[CV]  max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=100, score=0.084, total=   1.4s
[CV] max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=1000 
[CV]  max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=1000, score=0.124, total=  14.2s
[CV] max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=1000 
[CV]  max_depth=3, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=1000, score=0.251, total=

[Parallel(n_jobs=1)]: Done 120 out of 120 | elapsed: 52.4min finished


GridSearchCV(cv=5, error_score=nan,
             estimator=RandomForestRegressor(bootstrap=True, ccp_alpha=0.0,
                                             criterion='mse', max_depth=None,
                                             max_features='auto',
                                             max_leaf_nodes=None,
                                             max_samples=None,
                                             min_impurity_decrease=0.0,
                                             min_impurity_split=None,
                                             min_samples_leaf=1,
                                             min_samples_split=2,
                                             min_weight_fraction_leaf=0.0,
                                             n_estimators=100, n_jobs=None,
                                             oob_score=False, random_state=None,
                                             verbose=0, warm_start=False),
             iid='deprecated', n_jo

In [24]:
rf_grid.best_params_

{'max_depth': None,
 'max_features': 'auto',
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'n_estimators': 100}

In [26]:
from sklearn.svm import SVR


svr = GridSearchCV(SVR(kernel='rbf', gamma = 0.1),
                   param_grid={"C": [1e0, 1e1, 1e2, 1e3],
                               "gamma": np.logspace(-2, 2, 5)},
                   scoring = scorer,
                   cv = 5,
                   verbose = 2)

svr.fit(XX,yy)

Fitting 5 folds for each of 20 candidates, totalling 100 fits
[CV] C=1.0, gamma=0.01 ...............................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ................................ C=1.0, gamma=0.01, total=   6.0s
[CV] C=1.0, gamma=0.01 ...............................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    6.0s remaining:    0.0s


[CV] ................................ C=1.0, gamma=0.01, total=   5.9s
[CV] C=1.0, gamma=0.01 ...............................................
[CV] ................................ C=1.0, gamma=0.01, total=   6.0s
[CV] C=1.0, gamma=0.01 ...............................................
[CV] ................................ C=1.0, gamma=0.01, total=   6.0s
[CV] C=1.0, gamma=0.01 ...............................................
[CV] ................................ C=1.0, gamma=0.01, total=   5.9s
[CV] C=1.0, gamma=0.1 ................................................
[CV] ................................. C=1.0, gamma=0.1, total=   5.8s
[CV] C=1.0, gamma=0.1 ................................................
[CV] ................................. C=1.0, gamma=0.1, total=   5.9s
[CV] C=1.0, gamma=0.1 ................................................
[CV] ................................. C=1.0, gamma=0.1, total=   5.9s
[CV] C=1.0, gamma=0.1 ................................................
[CV] .

[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:  9.9min finished


GridSearchCV(cv=5, error_score=nan,
             estimator=SVR(C=1.0, cache_size=200, coef0=0.0, degree=3,
                           epsilon=0.1, gamma=0.1, kernel='rbf', max_iter=-1,
                           shrinking=True, tol=0.001, verbose=False),
             iid='deprecated', n_jobs=None,
             param_grid={'C': [1.0, 10.0, 100.0, 1000.0],
                         'gamma': array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02])},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=make_scorer(mean_squared_error, greater_is_better=False),
             verbose=2)

In [27]:
svr_best = svr.best_estimator_

pca = PCA(n_components = 0.95, svd_solver = 'full')

y_train_pred = []
y_test_pred = []

for train_index, test_index in tqdm(kf.split(XX), total = num_of_splits, unit = 'iteration'):

  X_train_std, X_test_std = XX[train_index,:], XX[test_index,:]
  y_train, y_test = yy.iloc[train_index], yy.iloc[test_index]

  pca.fit(XX)
  
  X_train_pca = pca.transform(X_train_std)
  X_test_pca = pca.transform(X_test_std)
  
  svr_best.fit(X_train_pca,y_train)
  
  y_train_pred.append(mean_squared_error(y_train,svr_best.predict(X_train_pca), squared = False))
  y_test_pred.append(mean_squared_error(y_test,svr_best.predict(X_test_pca), squared = False))

100%|██████████| 10/10 [00:50<00:00,  5.10s/iteration]


In [28]:
np.mean(y_train_pred)

7.57190670704602

In [29]:
np.mean(y_test_pred)

8.408145494162543